x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncSearchResponse.java

```diff
@@ -30,7 +30,7 @@
     @Nullable
     private final SearchResponse searchResponse;
     @Nullable
-    private final Exception error;
+    private final ElasticsearchException error;
```

Comment:
I don't think you can make that change without breaking BWC ? You'd need to wrap the exception if we read from an earlier version ?

Reply:
Although I think we should stick to an  `Exception` here to keep things simple

Reply:
I thought I have checked and this is fine. I haven't changed how the exception gets serialized? it was in fact always an elasticsearch exception before I think? I will check again.

Reply:
I am double checking: this change is not required but I think it simplifies things (I could also simplify the status method which I missed before) yet I agree that we have to make sure it does not break anything. 

The failure in MutableSearchResponse has always been an ElasticsearchException , and it used to be the only exception that gets passed in when building an AsyncSearchResponse. With this change we can also have a reduce exception, but that is still an ElasticsearchException. So, effectively, ElasticsearchException is the only exception that AsyncSearchResponse will ever hold. I think as long as we don't modify how we serialize it over the wire (by removing the type of exception because we already know its type) we should be ok? Is there anything I am missing that could cause a bw comp breakage?

Reply:
We had issues there before. The `ElasticsearchException` can be serialized over the wire and deserialized as another exception if it is not registered so I'd prefer that we keep `Exception` for now. I am not sure why do you think it simplifies things ?

Reply:
I think it simplifies things because it is super confusing to declare an exception when what we carry is always ElasticsearchException, and it simplifies returning the correct status, no guessing needed. I do get nervous though about the cast in StreamInput#readException, it is trappy and I was also wondering if this is not too risky. I reverted this bit, but I still don't get what it would break :) Possibly though it is wise to keep things as-is because async responses are stored in the index using the wire format which makes things tricky when it comes to bw comp.

Reply:
Yep the casting is trappy so +1 to keep it as is for the moment. We can change the way exceptions are handled in a follow up but that should be only for new versions imo.